### DISCUSSION TITLE
Help needed in understanding security concepts in EdgeX

### DISCUSSION THREAD

**Discussion by sudhamani-hcl** [edgexfoundry/community#93] [Q&A] [answered]
Hi Team,

We are trying to understand EdgeX security. But the information is there in bits and pieces. Could you please let us know the below queries? and also please share reference links & examples if any.

1) How to store redis username & password in vault? Which service is doing it and what configuration developer needs to do?
2) How to securely retrieve the credentials from Vault?
3) How to authenticate with vault? Is it enough to mention the secret key in MessageQueue section in configuration file?
4) We understand that secrets-token.json will be generated internally? Is it correct?
5) As a developer, How can we configure ACL?

Any help would be appreciated!!

Thanks in advance,
Sudhamani.

**Comment by bnevis-i** [comment]
> 1. How to store redis username & password in vault? Which service is doing it and what configuration developer needs to do?

security-bootstrapper contains startup scripts that are injected into 3rd party containers to add edgex-specific initialization logic.
security-secretstore-setup creates the redis password, and seeds it to the secretstore.
startup scripts injected into the redis container retrieve the redis password and write a redis config

> 2. How to securely retrieve the credentials from Vault?

security-secretstore-setup provides each service with its own unique vault token, the service then reads the secret out of secrets/edgex/(servicekey)/redisdb using the token

> 3. How to authenticate with vault? Is it enough to mention the secret key in MessageQueue section in configuration file?

authentication with vault is done via the vault token that is provided to each microservice (by default to /tmp/edgex/secrets/(servicekey)/secrets-token.json.  the secret name in the messagequeue section of the configuration file refers to the redis or message-bus credential in the service's slice of the secret store.

go-mod-bootstrap creates a secretprovider interface and injects it into the dependency injection container (dic).  the preinitialized object is already seeded with the vault token and service key specific to your service to be able to locate service-specific secrets.

> 4. We understand that secrets-token.json will be generated internally? Is it correct?

yes, it is generated by security-secretstore-setup (actually, by security-file-token-provider) at startup.  There is an extension mechanism provided to get the core logic to seed additional services: https://docs.edgexfoundry.org/3.0/security/Ch-Configuring-Add-On-Services/

> 5. As a developer, How can we configure ACL?

Whose ACL?  Vault, Redis, Consul are all ACL-based.

For Redis, we write the config files in this code:
https://github.com/edgexfoundry/edgex-go/blob/cc94d1f8a87c108527927d6732c4292434c72bde/internal/security/bootstrapper/helper/redis_conf.go

Unfortunately, to customize it you will need to fork security-bootstrapper.

We would of course take a contribution take a template file input that you could inject into the container where EdgeX would just replace the relevant section with a pre-seeded ACL.

**Reply by sudhamani-hcl** [reply]
> > 1. How to store redis username & password in vault? Which service is doing it and what configuration developer needs to do?
> 
> security-bootstrapper contains startup scripts that are injected into 3rd party containers to add edgex-specific initialization logic. security-secretstore-setup creates the redis password, and seeds it to the secretstore. startup scripts injected into the redis container retrieve the redis password and write a redis config
>>Thanks for the response Bryon. I still have not get information related to configuration. 
Is there any configuration user has to do related to redis username and password?
And also, I wanted to run EdgeX services with security enabled without using docker/snap. For this is there any reference document?
 
> > 2. How to securely retrieve the credentials from Vault?
> 
> security-secretstore-setup provides each service with its own unique vault token, the service then reads the secret out of secrets/edgex/(servicekey)/redisdb using the token
> 
> > 3. How to authenticate with vault? Is it enough to mention the secret key in MessageQueue section in configuration file?
> 
> authentication with vault is done via the vault token that is provided to each microservice (by default to /tmp/edgex/secrets/(servicekey)/secrets-token.json. the secret name in the messagequeue section of the configuration file refers to the redis or message-bus credential in the service's slice of the secret store.
> 
> go-mod-bootstrap creates a secretprovider interface and injects it into the dependency injection container (dic). the preinitialized object is already seeded with the vault token and service key specific to your service to be able to locate service-specific secrets.
> 
> > 4. We understand that secrets-token.json will be generated internally? Is it correct?
> 
> yes, it is generated by security-secretstore-setup (actually, by security-file-token-provider) at startup. There is an extension mechanism provided to get the core logic to seed additional services: https://docs.edgexfoundry.org/3.0/security/Ch-Configuring-Add-On-Services/
> 
> > 5. As a developer, How can we configure ACL?
> 
> Whose ACL? Vault, Redis, Consul are all ACL-based.
> 
> For Redis, we write the config files in this code: https://github.com/edgexfoundry/edgex-go/blob/cc94d1f8a87c108527927d6732c4292434c72bde/internal/security/bootstrapper/helper/redis_conf.go
> 
> Unfortunately, to customize it you will need to fork security-bootstrapper.
> 
> We would of course take a contribution take a template file input that you could inject into the container where EdgeX would just replace the relevant section with a pre-seeded ACL.
>> Is there any reference document for customizing ACL configuration? And also, What kind of secure communication happens b/n any service and message bus?

**Reply by bnevis-i** [reply]
> > > Is there any configuration user has to do related to redis username and password?

No.  It's all set up with a random password the first time you boot EdgeX.  Cleaning the docker volumes will do a full reset.

> > > And also, I wanted to run EdgeX services with security enabled without using docker/snap. For this is there any reference document?

No, there is not.  You are on your own.

> > > Is there any reference document for customizing ACL configuration? 

Consult the relevant Vault, Consul, Redis open source documentation.

> And also, What kind of secure communication happens b/n any service and message bus?

I am not the best person to answer.

**Reply by sudhamani-hcl** [reply]
Thank you Bryon for the comments.

We have one final question:
Does any of the EdgeX security services need hardware security features such as TPM? say for example true random number generation?

Thanks,
Sudhamani

**Reply by bnevis-i** [reply, marked as answer]
> We have one final question: Does any of the EdgeX security services need hardware security features such as TPM? say for example true random number generation?

One could argue that all security features start at the hardware, be it the instruction set architecture, encrypted memory, secure boot, et cetera.  As a middleware component, EdgeX assumes that the underlying platform components, hardware, hypervisors, operating system, container orchestration, etc. are all properly configured to operate in a secure mode.  You should evaluate the question above in that context.

Also, most of EdgeX is written in golang, which may use certain crypto acceleration CPU instructions, and uses the [getrandom(2) system call](https://pkg.go.dev/crypto/rand) and thus not directly under the control of the EdgeX stack.  That said, none of the EdgeX services -require- hardware security features to operate.  But if you have the desire to use some hardware security features, there are a few hooks:

- Enterprise Vault (vs the open source one) supports PKCS11 crypto modules to protect its sealing key -- you would have to fork EdgeX to be able to handle this.
- There is a [feature](https://docs.edgexfoundry.org/3.0/threat-models/secret-store/vault_master_key_encryption/) that allows one to pass in an encryption seed for the Vault (open source version) master key.  This key could be retrieved from security hardware, such as a TPM.
- EdgeX's delayed start services support runs on a software component called SPIFFE/SPIRE, which has hooks for [node attestation based on TPM2-based keys](https://github.com/spiffe/spire/blob/v1.6.3/doc/plugin_server_nodeattestor_tpm_devid.md)
- In EdgeX 3.0 (to be released), the nginx component supports PKCS11-based TLS keys

In any case, creating a direct software dependency on a hardware security feature would require a custom adaptation of EdgeX. 

You should threat model and analyze your full solution, of which EdgeX will just be one component.

You should also follow the Security WG's progress <https://wiki.edgexfoundry.org/display/FA/Security+Working+Group> to appraise yourself on the current security challenges that EdgeX is facing, such as fully authenticating all the top level APIs, input fuzzing, et cetera.  There is an inexhaustable list of possible improvements, and help is needed.

**Reply by sudhamani-hcl** [reply]
Thanks Bryon for the detailed response.
